### PR TITLE
Add commands to check and fix database corruption

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
- Improve the ACLK sync thread shutdown sequence
  -  Sets a `shutdown in progress` flag so that no more commands will be accepted and the current queue of the commands will be ignored

- Adds a -W parameter to check database for corruption and also attempt to fix it. 
- Adds a -W parameter to compact the database
- During startup if a database corruption is detected, it will attempt to fix it automatically (only applies
  to the dimension and chart tables)

Parameters added (in `netdata -h`) 
```
 -W check-sqlite-database        Check metadata database integrity and exit.
 -W fix-sqlite-database          Check metadata database integrity, fix if needed and exit.
 -W compact-sqlite-database      Reclaim metadata database unused space and exit.
```
##### Component Name
aclk sync, database

##### Test Plan
- Stop the agent and execute in turn
  - To simply check the database execute `netdata -W check-sqlite-database`
  - To check and fix if needed the database execute `netdata -W fix-sqlite-database`
  - To compact the database execute `netdata -W compact-sqlite-database`

A sample output of a corrupted database would look like this when `-W check-sqlite-database` is executed

```
2021-11-24 23:28:28: netdata INFO  : MAIN : Running database check on /home/stelios/netdata-staging/netdata/var/cache/netdata/netdata-meta.db
2021-11-24 23:28:28: netdata INFO  : MAIN : Checking table chart
2021-11-24 23:28:28: netdata INFO  : MAIN : ---> row 4300 missing from index sqlite_autoindex_chart_1
2021-11-24 23:28:30: netdata INFO  : MAIN : ---> row 439125 missing from index sqlite_autoindex_chart_1
2021-11-24 23:28:30: netdata INFO  : MAIN : ---> non-unique entry in index sqlite_autoindex_chart_1
2021-11-24 23:28:31: netdata ERROR : MAIN : Errors reported -- run with -W fix-sqlite-database
2021-11-24 23:28:31: netdata INFO  : MAIN : Checking table dimension
2021-11-24 23:28:39: netdata INFO  : MAIN : ---> row 1350373 missing from index ind_d1
2021-11-24 23:28:41: netdata INFO  : MAIN : ---> row 1744253 missing from index ind_d1
2021-11-24 23:28:42: netdata ERROR : MAIN : Errors reported -- run with -W fix-sqlite-database
```

### CODE REVIEW

**Comment by MrZammler** [comment]
Just a small comment, could `database` cause some confusion as to what it refers to? `sqlite database` or `dbengine database` ?

**Comment by stelfrag** [comment]
> Just a small comment, could `database` cause some confusion as to what it refers to? `sqlite database` or `dbengine database` ?

Good point, I will change the commands to 

- check-sqlite-database
- fix-sqlite-database
- compact-sqlite-database

File: database/sqlite/sqlite_functions.c [hunk 1]
```diff
@@ -180,18 +186,152 @@ void store_active_dimension(uuid_t *dimension_uuid)
...
+        snprintfz(wstr, 254, "PRAGMA integrity_check(%s);", table);
+    }
+    else {
+        info("Checking entire database");
+        strcpy(wstr,"PRAGMA integrity_check;");
+    }
+
+    int rc = sqlite3_exec(db_meta, wstr, check_table_integrity_cb, (void *) &status, &err_msg);
+    if (rc != SQLITE_OK) {
+        error_report("SQLite error during database integrity check for table %s, rc = %d (%s)", table, rc, err_msg);
```
Comment (MrZammler) [inline database/sqlite/sqlite_functions.c:217]: In case `table` is null, should there here be a check for it?

**Comment by MrZammler** [comment]
A general question:

Should there be a mechanism to declare a database as beyond-repair, perhaps then delete it, create a new one and allow Netdata to start eventually?

File: database/sqlite/sqlite_functions.c (hunk 1 above)
Comment (stelfrag) [reply database/sqlite/sqlite_functions.c]: Fix pushed

**Comment by stelfrag** [comment]
> A general question:
> 
> Should there be a mechanism to declare a database as beyond-repair, perhaps then delete it, create a new one and allow Netdata to start eventually?

It is a good idea but there needs to be a backup plan (literally a way to backup a last known usage database file to revert to)  as deleting the database, affects collected metrics (they become inaccessible)

**Review by underhood — APPROVED** [review]
I would perhaps rename all to start with sqlite e.g.
sqlite-check
sqlite-fix
sqlite-compact

**Comment by stelfrag** [comment]
> I would perhaps rename all to start with sqlite e.g. sqlite-check sqlite-fix sqlite-compact

yes simpler and shorter -- pushed commit

**Review by underhood — APPROVED** [review]

**Review by MrZammler — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]
I tested new options when netdata was stopped or running, and for both cases I got : 

```bash
bash-5.1# netdata -W sqlite-check  
2021-11-26 16:10:35: netdata INFO  : MAIN : Running database check on /var/cache/netdata/netdata-meta.db
2021-11-26 16:10:35: netdata INFO  : MAIN : Checking table chart
2021-11-26 16:10:35: netdata INFO  : MAIN : ---> ok
2021-11-26 16:10:35: netdata INFO  : MAIN : Checking table dimension
2021-11-26 16:10:36: netdata INFO  : MAIN : ---> ok
2021-11-26 16:10:36: netdata INFO  : MAIN : Checking entire database
2021-11-26 16:10:37: netdata INFO  : MAIN : ---> ok
bash-5.1# netdata -W sqlite-fix
2021-11-26 16:10:41: netdata INFO  : MAIN : Running database check on /var/cache/netdata/netdata-meta.db
2021-11-26 16:10:41: netdata INFO  : MAIN : Checking table chart
2021-11-26 16:10:41: netdata INFO  : MAIN : ---> ok
2021-11-26 16:10:41: netdata INFO  : MAIN : Checking table dimension
2021-11-26 16:10:41: netdata INFO  : MAIN : ---> ok
2021-11-26 16:10:41: netdata INFO  : MAIN : Checking entire database
2021-11-26 16:10:42: netdata INFO  : MAIN : ---> ok
bash-5.1# netdata -W sqlite-compact
2021-11-26 16:10:46: netdata INFO  : MAIN : Reclaiming space of /var/cache/netdata/netdata-meta.db
```
LGTM!

**Pull request merged**